### ISSUE TITLE
local ancestry of samples

### ISSUE THREAD

**Issue by rwaples** [jeromekelleher/msprime#371]
I'm not sure if this is a feature request or just a question. 

I am looking at the local genetic ancestry of admixed samples.  That is - what is the ancestral origin of  each genomic segment in each sample.  A common way to simulate data for this scenarios is to take  haplotypes for each population (maybe generated with a coalescent simulator) and then feed the haplotypes to a forward simulation to address the actual admixture (example: [admix-simu](https://github.com/williamslab/admix-simu)).  

Can all this be accomplished within mspime, without the need for the forward sim step? 
In terms of a coalescent simulation with admixture, I think this would boil down to specifying a time in the past and then annotating each sample with the population id of the ancestral node at the given time for each genomic interval.  More generally, given a sample and a time in the past, for each interval get the ancestor from that time.  This seems like it could be potentially useful information for a variety of analyses. 

I think I reconstruct this by tracing back relatives using ts.records() or from a combination of the migrations and edge tables, but I'm not sure that is the best way. 

Do you have any ideas on the best way to proceed, or do you suggest to stay with the forward simulations? 

msprime.__version__ = 0.4.1.dev696+g299ddc9

**Comment by jeromekelleher** [comment]
This is a great question, thanks @rwaples. I'm sure we can do this with msprime. I'm working on an msprime tutorial at the moment, so I'll try and work up an example that'll address this sort of question. I'll ping back here when its ready.

**Comment by rwaples** [comment]
That sounds great, thank you!

**Comment by jeromekelleher** [comment]
Thinking this through a bit, I realise I don't quite understand  the problem. 

How is this different to finding the population of the root node in each tree? The ancestral origin (among the entire sample) is where the MRCA of the sample occured... So, could we not just go through the trees, and get the population of each root node?

Or is this with respect to a subset of the sample?

**Comment by rwaples** [comment]
Thanks for thinking more about this.  I think what you describe above is a different question that doesn't relate to local ancestry in quite the same way, but I could be making this more confusing than it needs to be.  I’ll try to reformulate my description.

Here and above:
Sample = a single observed haplotype 

Local ancestry models consider that each sample is constructed from haplotypes inherited from a set of source populations that existed at some specific time in the past.  Practically, methods to infer local ancestry usually require reference panels that give example haplotypes from each source population. Then each sample is modeled as a mosaic of these reference haplotypes allowing for recombination. Often people care more about the source population of those haplotypes (modeling each sample as a combination of haplotypes from one or more source populations), and less about the specific haplotypes. 

Finding the population at the root node in each tree (as you suggest) is certainly interesting, but to me it doesn’t seem sufficient to address local ancestry.  It is missing the idea of defining the eligible populations at a specific time. Say there is a population 1, which splits in to populations 2 and 3 at time 10 in the past.  We have (possibly admixed) samples from the present and we want to consider the samples as mixtures of haplotypes from populations 2 and 3, as defined at time 5 in past. In this case, we never want to consider population 1 as an eligible source population, even though it could be the population with the MRCA.

For another take, see 

“Consider a sample and its ancestral population at some time T in the past, and suppose that we give ancestral individuals subpopulation labels, defining source populations. These labels are typically chosen to represent subgroups that have increased genetic homogeneity due to cultural or geographic reasons. Then a simple summary of the demographic trajectory of a sampled allele is the source population from which it originated. We say that an individual is “admixed” if it draws ancestry from multiple source populations—thus admixture is not an intrinsic property of individuals, but depends on our choice of labels and time T.”

From the introduction of:
Gravel S. Population Genetics Models of Local Ancestry. Genetics. 2012;191(2):607-619. doi:10.1534/genetics.112.139808.

**Comment by jeromekelleher** [comment]
Thanks for this @rwaples, I'll think about this some more. In the mean time, I've been writing some material for an msprime tutorial, which I was hoping would clear up some of these problems. Would you mind taking a look and seeing if it makes sense (and, if it helps with your application)?:

https://github.com/jeromekelleher/spg-chapter/blob/master/jupyter/introgression.ipynb

Thanks.

**Comment by rwaples** [comment]
I'm happy to, thanks.

**Comment by rwaples** [comment]
Thanks, that was a very helpful series of examples. I put up a very rough workup of what I was thinking here: [gist](https://gist.github.com/rwaples/10336129f75239465279dbfe163bf3c1) at the bottom of the introgression.ipynb you linked.  Essentially I created a function that takes a (sample, time, ts) and returns the population where each genomic region in the sample was at the given time. 

I added a very few comments in markdown as well - I was lost where a few of the numbers came from and the labels seemed switched on histogram.

**Comment by jeromekelleher** [comment]
This is really nice @rwaples,  powerful stuff! Thanks for the comments too, that's very helpful.

Two questions:
1) We're hoping to adapt this content into the online documentation somehow. Would it be (a) OK to use your example that you've added; and (b) would you be willing to help polish this the notebook up a bit?
2) Does this answer you original question (if so, can we close the issue)?

**Comment by rwaples** [comment]
Yes, I'm totally happy closing this issue, thanks for your help.  I'll let you close it.

And yes, please feel free to include this example.  I'm happy to help polish the notebook (and method) up a bit. 

We can pick the other issues up over there, but some immediate thoughts: 
- I think I could remove at least one of the loops in the code - we don't need to visit each tree twice. 
- Can we get by with just the example ts provided at the start?
- I'd also like to add some checks so that I'm sure I am getting the right answer in more complex scenarios.

Not sure what timeframe you have mind, but I'm not sure ill have a chance before the second half of February.  I will get back to this though. I need data like this to test a method looking at LD.

Thanks again,
Ryan
